packages/docregistry/src/context.ts

```diff
@@ -454,6 +456,62 @@ class Context<T extends DocumentRegistry.IModel> implements DocumentRegistry.ICo
     });
   }
 
+  /**
+   * Save a file, dealing with conflicts.
+   */
+  private _maybeSave(options: Partial<Contents.IModel>): Promise<Contents.IModel> {
+    let path = this._path;
+    // Make sure the file has not changed on disk.
+    let promise = this._manager.contents.get(path, { content: false });
+    return promise.then(model => {
+      if (this.isDisposed) {
+        return Promise.reject('Disposed');
+      }
+      // We want to check last_modified (disk) > last_modified (client)
+      // (our last save)
+      // In some cases the filesystem reports an inconsistent time,
+      // so we allow 0.5 seconds difference before complaining.
+      let modified = this.contentsModel && this.contentsModel.last_modified;
+      let tClient = new Date(modified);
+      let tDisk = new Date(model.last_modified);
+      if (modified && (tDisk.getTime() - tClient.getTime()) > 500) {  // 500 ms
+        return this._timeConflict(tClient, model, options);
+      }
+      return this._manager.contents.save(path, options);
+    }, (err) => {
+      return this._manager.contents.save(path, options);
```

Comment:
This implies that the only reason for `promise` to reject is non-existence of the file. At some future point, we'll probably want to catch more errors, *i.e.*, server error.